File: src/lasercut/face_classifier.py
"""Classify faces into bottom/walls and build adjacency graph."""

import math
from dataclasses import dataclass
from .step_loader import PlanarFace, EdgeData


@dataclass
class SharedEdge:
    """A shared edge between two faces."""
    face_a_id: int
    face_b_id: int
    edge_a: EdgeData
    edge_b: EdgeData
    midpoint: tuple[float, float, float]
    length: float


def _edge_length(edge: EdgeData) -> float:
    dx = edge.end[0] - edge.start[0]
    dy = edge.end[1] - edge.start[1]
    dz = edge.end[2] - edge.start[2]
    return math.sqrt(dx * dx + dy * dy + dz * dz)


def _points_close(p1: tuple, p2: tuple, tol: float) -> bool:
    return all(abs(a - b) < tol for a, b in zip(p1, p2))


def _edges_coincident(e1: EdgeData, e2: EdgeData, tol: float = 0.1) -> bool:
    """Check if two edges are geometrically coincident (same or reversed direction)."""
    if _points_close(e1.start, e2.start, tol) and _points_close(e1.end, e2.end, tol):
        return True
    if _points_close(e1.start, e2.end, tol) and _points_close(e1.end, e2.start, tol):
        return True
    if _points_close(e1.midpoint, e2.midpoint, tol):
        length1 = _edge_length(e1)
        length2 = _edge_length(e2)
        if abs(length1 - length2) < tol:
            return True
    return False


def _dot3(a: tuple, b: tuple) -> float:
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def find_shared_edges(faces: list[PlanarFace], tol: float = 0.1) -> list[SharedEdge]:
    """Find edges shared between pairs of faces."""
    shared = []
    for i, face_a in enumerate(faces):
        for j, face_b in enumerate(faces):
            if j <= i:
                continue
            for ea in face_a.outer_wire_edges:
                for eb in face_b.outer_wire_edges:
                    if _edges_coincident(ea, eb, tol):
                        length = _edge_length(ea)
                        if length > tol:
                            shared.append(SharedEdge(
                                face_a_id=face_a.face_id,
                                face_b_id=face_b.face_id,
                                edge_a=ea,
                                edge_b=eb,
                                midpoint=ea.midpoint,
                                length=length,
                            ))
    return shared


def build_adjacency(faces: list[PlanarFace], shared_edges: list[SharedEdge]) -> dict[int, list[tuple[int, SharedEdge]]]:
    """Build adjacency graph: face_id -> [(neighbor_id, shared_edge), ...]."""
    adj: dict[int, list[tuple[int, SharedEdge]]] = {f.face_id: [] for f in faces}
    for se in shared_edges:
        adj[se.face_a_id].append((se.face_b_id, se))
        adj[se.face_b_id].append((se.face_a_id, se))
    return adj


def _find_opposite_pairs(faces: list[PlanarFace], tol: float = 0.05) -> dict[int, int]:
    """Find pairs of faces with opposite normals and similar area (inner/outer of same panel).

    Returns mapping: face_id -> paired_face_id (only for the LARGER face of each pair).
    """
    pairs = {}
    used = set()

    # Sort by area descending to prefer pairing larger faces first
    sorted_faces = sorted(faces, key=lambda f: f.area, reverse=True)

    for i, fa in enumerate(sorted_faces):
        if fa.face_id in used:
            continue
        for j, fb in enumerate(sorted_faces):
            if j <= i or fb.face_id in used:
                continue
            # Check if normals are opposite
            dot = _dot3(fa.normal, fb.normal)
            if dot < -(1.0 - tol):
                # Normals are opposite; check area similarity (within 20%)
                area_ratio = min(fa.area, fb.area) / max(fa.area, fb.area)
                if area_ratio > 0.8:
                    # fa is larger (sorted by area desc), fb is smaller (inner face)
                    pairs[fa.face_id] = fb.face_id
                    used.add(fa.face_id)
                    used.add(fb.face_id)
                    break

    return pairs


def classify_faces(faces: list[PlanarFace], shared_edges: list[SharedEdge]) -> dict:
    """Classify faces into bottom plate and walls.

    Strategy:
    1. Find inner/outer face pairs (opposite normals, similar area)
    2. Take the larger (outer) face from each pair as a structural panel
    3. The largest structural panel is the bottom
    4. All other structural panels connected to the bottom (directly or via
       other panels) are walls

    Returns:
        Dictionary with keys:
        - 'bottom': PlanarFace (largest structural face)
        - 'walls': list of PlanarFace
        - 'other': list of PlanarFace
        - 'adjacency': adjacency graph
    """
    if not faces:
        return {'bottom': None, 'walls': [], 'other': [], 'adjacency': {}}

    adjacency = build_adjacency(faces, shared_edges)
    face_map = {f.face_id: f for f in faces}

    # Find inner/outer pairs
    pairs = _find_opposite_pairs(faces)
    outer_ids = set(pairs.keys())
    inner_ids = set(pairs.values())

    # Structural faces: outer faces from pairs + unpaired large faces
    structural_ids = set(outer_ids)
    for f in faces:
        if f.face_id not in outer_ids and f.face_id not in inner_ids:
            # Unpaired face — include if large enough (>5% of largest face area)
            max_area = max(ff.area for ff in faces)
            if f.area > max_area * 0.05:
                structural_ids.add(f.face_id)

    if not structural_ids:
        structural_ids = {f.face_id for f in faces}

    # Bottom: largest structural face
    bottom = max((face_map[fid] for fid in structural_ids), key=lambda f: f.area)

    # Find all walls: flood-fill from bottom through adjacency,
    # but only visiting other structural faces
    wall_ids = set()
    queue = [bottom.face_id]
    visited = {bottom.face_id}

    while queue:
        current = queue.pop(0)
        for neighbor_id, _ in adjacency.get(current, []):
            if neighbor_id in visited:
                continue
            if neighbor_id in structural_ids:
                wall_ids.add(neighbor_id)
                visited.add(neighbor_id)
                queue.append(neighbor_id)

    walls = [face_map[fid] for fid in wall_ids]
    classified_ids = {bottom.face_id} | wall_ids
    other = [f for f in faces if f.face_id not in classified_ids]

    return {
        'bottom': bottom,
        'walls': walls,
        'other': other,
        'adjacency': adjacency,
    }

File: lasercut-rs/src/face_classifier.rs
use std::collections::{HashMap, HashSet, VecDeque};

use crate::math_utils::{dist3, dot3, norm3, normalize3, sub3};
use crate::types::{EdgeData, FaceClassification, PlanarFace, SharedEdge};

fn edge_length(edge: &EdgeData) -> f64 {
    dist3(edge.start, edge.end)
}

fn points_close(p1: [f64; 3], p2: [f64; 3], tol: f64) -> bool {
    (p1[0] - p2[0]).abs() < tol && (p1[1] - p2[1]).abs() < tol && (p1[2] - p2[2]).abs() < tol
}

fn edges_coincident(e1: &EdgeData, e2: &EdgeData, tol: f64) -> bool {
    // Forward match
    if points_close(e1.start, e2.start, tol) && points_close(e1.end, e2.end, tol) {
        return true;
    }
    // Reverse match
    if points_close(e1.start, e2.end, tol) && points_close(e1.end, e2.start, tol) {
        return true;
    }
    // Midpoint + length match
    if points_close(e1.midpoint, e2.midpoint, tol) {
        let len1 = edge_length(e1);
        let len2 = edge_length(e2);
        if (len1 - len2).abs() < tol {
            return true;
        }
    }
    false
}

/// Find edges shared between pairs of faces.
pub fn find_shared_edges(faces: &[PlanarFace], tol: f64) -> Vec<SharedEdge> {
    let mut shared = Vec::new();

    for i in 0..faces.len() {
        for j in (i + 1)..faces.len() {
            for ea in &faces[i].outer_wire_edges {
                for eb in &faces[j].outer_wire_edges {
                    if edges_coincident(ea, eb, tol) {
                        let length = edge_length(ea);
                        if length > tol {
                            shared.push(SharedEdge {
                                face_a_id: faces[i].face_id,
                                face_b_id: faces[j].face_id,
                                edge_a: ea.clone(),
                                edge_b: eb.clone(),
                                midpoint: ea.midpoint,
                                length,
                            });
                        }
                    }
                }
            }
        }
    }

    shared
}

/// Build adjacency graph: face_id -> [(neighbor_id, shared_edge), ...].
pub fn build_adjacency(
    faces: &[PlanarFace],
    shared_edges: &[SharedEdge],
) -> HashMap<usize, Vec<(usize, SharedEdge)>> {
    let mut adj: HashMap<usize, Vec<(usize, SharedEdge)>> = HashMap::new();
    for f in faces {
        adj.entry(f.face_id).or_default();
    }
    for se in shared_edges {
        adj.entry(se.face_a_id)
            .or_default()
            .push((se.face_b_id, se.clone()));
        adj.entry(se.face_b_id)
            .or_default()
            .push((se.face_a_id, se.clone()));
    }
    adj
}

/// Find pairs of faces with opposite normals and similar area (inner/outer of same panel).
fn find_opposite_pairs(
    faces: &[PlanarFace],
    tol: f64,
    max_offset: f64,
    max_lateral_offset: f64,
) -> HashMap<usize, usize> {
    let mut pairs = HashMap::new();
    let mut used = HashSet::new();

    let mut sorted_faces: Vec<&PlanarFace> = faces.iter().collect();
    sorted_faces.sort_by(|a, b| b.area.partial_cmp(&a.area).unwrap());

    for (i, fa) in sorted_faces.iter().enumerate() {
        if used.contains(&fa.face_id) {
            continue;
        }

        let n_len = norm3(fa.normal);
        if n_len < 1e-12 {
            continue;
        }
        let n_hat = normalize3(fa.normal);

        let mut best_candidate: Option<(f64, usize)> = None;

        for (j, fb) in sorted_faces.iter().enumerate() {
            if j <= i || used.contains(&fb.face_id) {
                continue;
            }

            let dot = dot3(fa.normal, fb.normal);
            if dot < -(1.0 - tol) {
                let area_ratio = fa.area.min(fb.area) / fa.area.max(fb.area);
                if area_ratio > 0.8 {
                    let delta = sub3(fb.center, fa.center);
                    let normal_proj = dot3(delta, n_hat);
                    let normal_offset = normal_proj.abs();
                    let lateral = [
                        delta[0] - normal_proj * n_hat[0],
                        delta[1] - normal_proj * n_hat[1],
                        delta[2] - normal_proj * n_hat[2],
                    ];
                    let lateral_offset = norm3(lateral);

                    if normal_offset <= max_offset && lateral_offset <= max_lateral_offset {
                        let score = normal_offset + 0.1 * lateral_offset;
                        if best_candidate.is_none() || score < best_candidate.unwrap().0 {
                            best_candidate = Some((score, fb.face_id));
                        }
                    }
                }
            }
        }

        if let Some((_, match_id)) = best_candidate {
            pairs.insert(fa.face_id, match_id);
            used.insert(fa.face_id);
            used.insert(match_id);
        }
    }

    pairs
}

/// Classify faces into bottom plate and walls.
pub fn classify_faces(
    faces: &[PlanarFace],
    shared_edges: &[SharedEdge],
) -> FaceClassification {
    if faces.is_empty() {
        panic!("No faces to classify");
    }

    let adjacency = build_adjacency(faces, shared_edges);
    let face_map: HashMap<usize, &PlanarFace> = faces.iter().map(|f| (f.face_id, f)).collect();

    // Find inner/outer pairs
    let pairs = find_opposite_pairs(faces, 0.05, 20.0, 20.0);
    let outer_ids: HashSet<usize> = pairs.keys().copied().collect();
    let inner_ids: HashSet<usize> = pairs.values().copied().collect();

    // Structural faces: outer faces from pairs + unpaired large faces
    let mut structural_ids: HashSet<usize> = outer_ids.clone();
    let max_area = faces.iter().map(|f| f.area).fold(0.0_f64, f64::max);

    for f in faces {
        if !outer_ids.contains(&f.face_id) && !inner_ids.contains(&f.face_id) {
            if f.area > max_area * 0.05 {
                structural_ids.insert(f.face_id);
            }
        }
    }

    if structural_ids.is_empty() {
        structural_ids = faces.iter().map(|f| f.face_id).collect();
    }

    // Bottom: largest structural face
    let bottom_id = structural_ids
        .iter()
        .max_by(|a, b| {
            face_map[a]
                .area
                .partial_cmp(&face_map[b].area)
                .unwrap()
        })
        .copied()
        .unwrap();

    // Walls: flood-fill from bottom through adjacency.
    // Traverse ALL faces (including inner faces) but only collect structural faces as walls.
    // This ensures we reach structural faces that are only adjacent to inner faces
    // (e.g. back wall connected through inner surfaces of side walls).
    let mut wall_ids = HashSet::new();
    let mut visited = HashSet::new();
    visited.insert(bottom_id);
    let mut queue = VecDeque::new();
    queue.push_back(bottom_id);

    while let Some(current) = queue.pop_front() {
        if let Some(neighbors) = adjacency.get(&current) {
            for (neighbor_id, _) in neighbors {
                if visited.contains(neighbor_id) {
                    continue;
                }
                visited.insert(*neighbor_id);
                queue.push_back(*neighbor_id);
                if structural_ids.contains(neighbor_id) {
                    wall_ids.insert(*neighbor_id);
                }
            }
        }
    }

    let bottom = face_map[&bottom_id].clone();
    let walls: Vec<PlanarFace> = wall_ids.iter().map(|id| face_map[id].clone()).collect();
    let classified_ids: HashSet<usize> = std::iter::once(bottom_id).chain(wall_ids).collect();
    let other: Vec<PlanarFace> = faces
        .iter()
        .filter(|f| !classified_ids.contains(&f.face_id))
        .cloned()
        .collect();

    FaceClassification {
        bottom,
        walls,
        other,
        adjacency,
    }
}
